### PR TITLE
feat: Run a `SlidingSync` when retrieving notifications to get more information

### PR DESCRIPTION
**Pull request by bnjbvr**
This implements what's been suggested in #2179: a short, non-e2ee, sliding sync is run in the notification client, so as to retrieve the notification as well as extra context required for display in the notification. This requires that the client is built with an in-memory state store (https://github.com/matrix-org/matrix-rust-sdk/pull/2241), or it may corrupt the state store in unexpected ways.

Address https://github.com/matrix-org/matrix-rust-sdk/issues/2179.

For ElementX Android, we might use a different solution based on the `/context` endpoint, using lazy loading of the member information.

### CODE REVIEW

**Comment by Velin92** [comment]
This branch seems to make the iOS app crash, when the first sync happens (tested on an actual iOS device):
```
2023-07-12T13:22:31.886053Z DEBUG sync_stream:sync_once{pos="2"}:send{homeserver="https://slidingsync.lab.matrix.org/" server_versions=[V1_0, V1_1, V1_2, V1_3, V1_4, V1_5] config=RequestConfig { timeout: 60s } request_id="REQ-8" request_size="129 B" request_body=b"{\"conn_id\":\"room-list\",\"lists\":{\"all_rooms\":{\"ranges\":[[0,4]]},\"invites\":{\"ranges\":[[0,0]]},\"visible_rooms\":{\"ranges\":[[0,19]]}}}" path="/_matrix/client/unstable/org.matrix.msc3575/sync?pos=2&timeout=30000"}: matrix_sdk::http_client: crates/matrix-sdk/src/http_client/mod.rs:198: Sending request
2023-07-12T13:22:31.886284Z  INFO root: elementx: ClientProxy.swift:423: Received room list update: running
   0:        0x106385070 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h1f74b598feff883f
   1:        0x1063d06a8 - core::fmt::write::ha43a0a9828d2ac77
   2:        0x10637ad6c - std::io::Write::write_fmt::h0b366bb6709b84c6
   3:        0x106384ec4 - std::sys_common::backtrace::print::hc0cf930553f44a1e
   4:        0x1063876b0 - std::panicking::default_hook::{{closure}}::hbd5bdc4212351e96
   5:        0x1063873e8 - std::panicking::default_hook::h1ffa666ddeaab943
   6:        0x106387d74 - std::panicking::rust_panic_with_hook::h6e42f6c8e7b4a238
   7:        0x106387be0 - std::panicking::begin_panic_handler::{{closure}}::hafa4b8d86403be24
   8:        0x106385450 - std::sys_common::backtrace::__rust_end_short_backtrace::h4d42fe977ebdd26f
   9:        0x106387954 - _rust_begin_unwind
  10:        0x10645c70c - core::panicking::panic_fmt::he823a3417d454303
  11:        0x106087898 - tokio::runtime::scheduler::Handle::current::h39ad9628798ff1ab
  12:        0x1056190ac - tokio::runtime::blocking::pool::spawn_blocking::h50426e07eaf46336
  13:        0x10584e1cc - core::ptr::drop_in_place<deadpool_sync::SyncWrapper<rusqlite::Connection>>::h15ee24813424f6db
  14:        0x105851818 - core::ptr::drop_in_place<deadpool::managed::UnreadyObject<deadpool_sqlite::Manager>>::h4c9bdf236a1471d1
  15:        0x1057eb294 - matrix_sdk_sqlite::state_store::SqliteStateStore::acquire::{{closure}}::h148f8dda1c6f20cc
  16:        0x1057ee98c - <matrix_sdk_sqlite::state_store::SqliteStateStore as matrix_sdk_base::store::traits::StateStore>::get_state_events_for_keys::{{closure}}::h7c519da406c2c8ab
  17:        0x1057edd58 - <matrix_sdk_sqlite::state_store::SqliteStateStore as matrix_sdk_base::store::traits::StateStore>::get_state_event::{{closure}}::ha4c62cc58790ac74
  18:        0x104882a98 - <matrix_sdk_base::store::traits::EraseStateStoreError<T> as matrix_sdk_base::store::traits::StateStore>::get_state_event::{{closure}}::hd951aae08b2c044e
  19:        0x1052eae04 - matrix_sdk_base::store::traits::StateStoreExt::get_state_event_static_for_key::{{closure}}::hd7dea4d20fe63255
  20:        0x1052ea6d0 - matrix_sdk_base::store::traits::StateStoreExt::get_member_event::{{closure}}::hccdcbfa59b351174
  21:        0x104fa2d84 - matrix_sdk::room::common::Common::get_member_no_sync::{{closure}}::h4e0b463eb2b4c3bf
  22:        0x104fb54f4 - <matrix_sdk::room::common::Common as matrix_sdk_ui::timeline::traits::RoomDataProvider>::profile::{{closure}}::ha27426f6e5ddea7a
  23:        0x10519f5dc - <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll::h222f91305a7f7bed
  24:        0x1050ac494 - <matrix_sdk::sliding_sync::room::SlidingSyncRoom as matrix_sdk_ui::timeline::sliding_sync_ext::SlidingSyncRoomExt>::latest_timeline_item::{{closure}}::hbb4ea691447db5b8
  25:        0x1048ab890 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::had488fe1c2eaf246
  26:        0x104cd7320 - uniffi_core::ffi::rustcalls::rust_call_with_out_status::h2a73b31540bd6807
  27:        0x1041c79d4 - uniffi_core::ffi::rustfuture::RustFuture<F,T,UT>::wake_callback::h939da371a1031324
  28:        0x1037cb0f4 - _$s13MatrixRustSDK29uniffiForeignExecutorCallback33_D13DA838D267D18EF7DBCAF4C15FEA09LL14executorHandle7delayMs8rustTask8taskDataySi_s6UInt32VySVSgXCSgAJtFyyYaYbcfU_TY0_
  29:        0x1a7252948 - <unknown>
  30:        0x1a7253868 - <unknown>
  31:        0x10c39643c - __dispatch_root_queue_drain
  32:        0x10c396e34 - __dispatch_worker_thread2
  33:        0x1fcc20da0 - <unknown>
thread '<unnamed>' panicked at 'panic in a function that cannot unwind', library/core/src/panicking.rs:126:5
stack backtrace:
   0:        0x106385070 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h1f74b598feff883f
   1:        0x1063d06a8 - core::fmt::write::ha43a0a9828d2ac77
   2:        0x10637ad6c - std::io::Write::write_fmt::h0b366bb6709b84c6
   3:        0x106384ec4 - std::sys_common::backtrace::print::hc0cf930553f44a1e
   4:        0x1063876b0 - std::panicking::default_hook::{{closure}}::hbd5bdc4212351e96
   5:        0x1063873e8 - std::panicking::default_hook::h1ffa666ddeaab943
   6:        0x106387d74 - std::panicking::rust_panic_with_hook::h6e42f6c8e7b4a238
   7:        0x106387ba4 - std::panicking::begin_panic_handler::{{closure}}::hafa4b8d86403be24
   8:        0x106385450 - std::sys_common::backtrace::__rust_end_short_backtrace::h4d42fe977ebdd26f
   9:        0x106387954 - _rust_begin_unwind
  10:        0x10645c73c - core::panicking::panic_nounwind_fmt::hcbaccdfcb27e0d35
  11:        0x10645c7b8 - core::panicking::panic_nounwind::hb54b0c3ed6ba0bab
  12:        0x10645c8ac - core::panicking::panic_cannot_unwind::h4bc393e3eb5c9cc8
  13:        0x1057e9aec - matrix_sdk_sqlite::state_store::SqliteStateStore::acquire::{{closure}}::h148f8dda1c6f20cc
  14:        0x1057ee98c - <matrix_sdk_sqlite::state_store::SqliteStateStore as matrix_sdk_base::store::traits::StateStore>::get_state_events_for_keys::{{closure}}::h7c519da406c2c8ab
  15:        0x1057edd58 - <matrix_sdk_sqlite::state_store::SqliteStateStore as matrix_sdk_base::store::traits::StateStore>::get_state_event::{{closure}}::ha4c62cc58790ac74
  16:        0x104882a98 - <matrix_sdk_base::store::traits::EraseStateStoreError<T> as matrix_sdk_base::store::traits::StateStore>::get_state_event::{{closure}}::hd951aae08b2c044e
  17:        0x1052eae04 - matrix_sdk_base::store::traits::StateStoreExt::get_state_event_static_for_key::{{closure}}::hd7dea4d20fe63255
  18:        0x1052ea6d0 - matrix_sdk_base::store::traits::StateStoreExt::get_member_event::{{closure}}::hccdcbfa59b351174
  19:        0x104fa2d84 - matrix_sdk::room::common::Common::get_member_no_sync::{{closure}}::h4e0b463eb2b4c3bf
  20:        0x104fb54f4 - <matrix_sdk::room::common::Common as matrix_sdk_ui::timeline::traits::RoomDataProvider>::profile::{{closure}}::ha27426f6e5ddea7a
  21:        0x10519f5dc - <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll::h222f91305a7f7bed
  22:        0x1050ac494 - <matrix_sdk::sliding_sync::room::SlidingSyncRoom as matrix_sdk_ui::timeline::sliding_sync_ext::SlidingSyncRoomExt>::latest_timeline_item::{{closure}}::hbb4ea691447db5b8
  23:        0x1048ab890 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::had488fe1c2eaf246
  24:        0x104cd7320 - uniffi_core::ffi::rustcalls::rust_call_with_out_status::h2a73b31540bd6807
  25:        0x1041c79d4 - uniffi_core::ffi::rustfuture::RustFuture<F,T,UT>::wake_callback::h939da371a1031324
  26:        0x1037cb0f4 - _$s13MatrixRustSDK29uniffiForeignExecutorCallback33_D13DA838D267D18EF7DBCAF4C15FEA09LL14executorHandle7delayMs8rustTask8taskDataySi_s6UInt32VySVSgXCSgAJtFyyYaYbcfU_TY0_
  27:        0x1a7252948 - <unknown>
  28:        0x1a7253868 - <unknown>
  29:        0x10c39643c - __dispatch_root_queue_drain
  30:        0x10c396e34 - __dispatch_worker_thread2
  31:        0x1fcc20da0 - <unknown>
thread caused non-unwinding panic. aborting.
```

This kinda reminds me the issue that happened when building in debug mode made uniFFI crash... I want to check if this happens in release mode too.

**Comment by Velin92** [comment]
Now everything works... aside from invites:
2```
023-07-13T13:27:52.319667Z  INFO root: nse: NSELogger.swift:88: [NSE][0x0000000153d56210][0x0000000155a908d0] Memory: footprint: 16.45 MB - available: 7.55 MB
2023-07-13T13:27:52.320102Z  INFO root: nse: NotificationServiceExtension.swift:51: [NSE][0x0000000153d56210][0x0000000155a908d0] #########################################
2023-07-13T13:27:52.320619Z  INFO root: nse: NotificationServiceExtension.swift:52: [NSE][0x0000000153d56210][0x0000000155a908d0] Payload came: [AnyHashable("pusher_notification_client_identifier"): 87b2a68295015bd4da30fba560ceabe94d86ad4ffd83876b74e69837616beb42, AnyHashable("room_id"): !vxNTUVPhUbFSlaIzTm:element.io, AnyHashable("unread_count"): 3, AnyHashable("event_id"): $yiB0FSbd241u4XGEyF3Tjgw37I6mhBpkVmAaGEYC5PU, AnyHashable("aps"): {
    alert =     {
        "loc-args" =         (
        );
        "loc-key" = Notification;
    };
    "mutable-content" = 1;
}]
2023-07-13T13:27:52.321109Z  INFO root: nse: NotificationServiceExtension.swift:71: [NSE][0x0000000153d56210][0x0000000155a908d0] run with roomId: !vxNTUVPhUbFSlaIzTm:element.io, eventId: $yiB0FSbd241u4XGEyF3Tjgw37I6mhBpkVmAaGEYC5PU
2023-07-13T13:27:52.324570Z DEBUG root:build{homeserver=ServerName("matrix.org")}:send{homeserver="https://matrix.org" server_versions=[V1_0] config=RequestConfig { timeout: 30s, retry_limit: 3 } request_id="REQ-0" request_size="0 B" path="/.well-known/matrix/client"}: matrix_sdk::http_client: crates/matrix-sdk/src/http_client/mod.rs:198: Sending request
2023-07-13T13:27:52.498570Z DEBUG root:build{homeserver=ServerName("matrix.org")}:send{homeserver="https://matrix.org" server_versions=[V1_0] config=RequestConfig { timeout: 30s, retry_limit: 3 } request_id="REQ-0" request_size="0 B" path="/.well-known/matrix/client" status=200 response_size="254 B"}: matrix_sdk::http_client: crates/matrix-sdk/src/http_client/mod.rs:204: Got response
2023-07-13T13:27:52.507053Z DEBUG root:restore_session:restore_session:with_store{user_id="@mauro.devtest:matrix.org" device_id="NVCMHPHFLM" ed25519_key=yr7crqTG26cc/2UnxnKn9QYjTDCwDmI4o4NYRARzzig curve25519_key=5YZha0ErXyHbfGFfgFo+zNYkZvN1UulkFtSWY8Ke9wI}: matrix_sdk_crypto::machine: crates/matrix-sdk-crypto/src/machine.rs:273: Restored an Olm account
2023-07-13T13:27:52.507974Z DEBUG root:restore_session:restore_session:with_store{user_id="@mauro.devtest:matrix.org" device_id="NVCMHPHFLM" ed25519_key=yr7crqTG26cc/2UnxnKn9QYjTDCwDmI4o4NYRARzzig curve25519_key=5YZha0ErXyHbfGFfgFo+zNYkZvN1UulkFtSWY8Ke9wI}: matrix_sdk_crypto::machine: crates/matrix-sdk-crypto/src/machine.rs:310: Restored the cross signing identity master_key=Some("ed25519:vc6gJSuY/HkNpYT6VIhXLpEN5gVLWBKqmUp30UUPsK8")
2023-07-13T13:27:52.509814Z  INFO root: matrix_sdk_ui::notification_client: crates/matrix-sdk-ui/src/notification_client.rs:245: fetching notification event with a sliding sync
2023-07-13T13:27:52.509868Z TRACE root: matrix_sdk::sliding_sync::cache: crates/matrix-sdk/src/sliding_sync/cache.rs:202: Failed to find the `SlidingSync` object in the cache
2023-07-13T13:27:52.510020Z DEBUG root:sync_stream: matrix_sdk::sliding_sync: crates/matrix-sdk/src/sliding_sync/mod.rs:603: About to run the sync-loop self.inner.position=RwLock { data: SlidingSyncPositionMarkers { pos: None, delta_token: None }, poisoned: false, .. }
2023-07-13T13:27:52.510047Z DEBUG root:sync_stream: matrix_sdk::sliding_sync: crates/matrix-sdk/src/sliding_sync/mod.rs:611: Sync-loop is running self.inner.position=RwLock { data: SlidingSyncPositionMarkers { pos: None, delta_token: None }, poisoned: false, .. }
2023-07-13T13:27:52.510137Z DEBUG root:sync_stream:sync_once: matrix_sdk::sliding_sync: crates/matrix-sdk/src/sliding_sync/mod.rs:468: Sending the sliding sync request
2023-07-13T13:27:52.510156Z DEBUG root:sync_stream:sync_once: matrix_sdk::sliding_sync: crates/matrix-sdk/src/sliding_sync/mod.rs:528: Sliding Sync is sending the request (e2ee not enabled in this instance)
2023-07-13T13:27:52.510252Z DEBUG root:sync_stream:sync_once:send{homeserver="https://matrix-client.matrix.org/" server_versions=[V1_0] config=RequestConfig { timeout: 30s } request_id="REQ-1" request_size="0 B" path="/_matrix/client/versions"}: matrix_sdk::http_client: crates/matrix-sdk/src/http_client/mod.rs:198: Sending request
2023-07-13T13:27:52.718945Z DEBUG root:sync_stream:sync_once:send{homeserver="https://matrix-client.matrix.org/" server_versions=[V1_0] config=RequestConfig { timeout: 30s } request_id="REQ-1" request_size="0 B" path="/_matrix/client/versions" status=200 response_size="857 B"}: matrix_sdk::http_client: crates/matrix-sdk/src/http_client/mod.rs:204: Got response
2023-07-13T13:27:52.719179Z DEBUG root:sync_stream:sync_once:send{homeserver="https://slidingsync.lab.matrix.org/" server_versions=[V1_0, V1_1, V1_2, V1_3, V1_4, V1_5, V1_6] config=RequestConfig { timeout: 4s } request_id="REQ-2" request_size="387 B" request_body=b"{\"conn_id\":\"notifications\",\"txn_id\":\"7a806bae405940018aceae6d94a01b8a\",\"room_subscriptions\":{\"!vxNTUVPhUbFSlaIzTm:element.io\":{\"required_state\":[[\"m.room.avatar\",\"\"],[\"m.room.encryption\",\"\"],[\"m.room.member\",\"$LAZY\"],[\"m.room.member\",\"$ME\"],[\"m.room.canonical_alias\",\"\"],[\"m.room.name\",\"\"],[\"m.room.power_levels\",\"\"]],\"timeline_limit\":16}},\"extensions\":{\"account_data\":{\"enabled\":true}}}" path="/_matrix/client/unstable/org.matrix.msc3575/sync?timeout=3000"}: matrix_sdk::http_client: crates/matrix-sdk/src/http_client/mod.rs:198: Sending request
2023-07-13T13:27:54.113604Z DEBUG root:sync_stream:sync_once:send{homeserver="https://slidingsync.lab.matrix.org/" server_versions=[V1_0, V1_1, V1_2, V1_3, V1_4, V1_5, V1_6] config=RequestConfig { timeout: 4s } request_id="REQ-2" request_size="387 B" request_body=b"{\"conn_id\":\"notifications\",\"txn_id\":\"7a806bae405940018aceae6d94a01b8a\",\"room_subscriptions\":{\"!vxNTUVPhUbFSlaIzTm:element.io\":{\"required_state\":[[\"m.room.avatar\",\"\"],[\"m.room.encryption\",\"\"],[\"m.room.member\",\"$LAZY\"],[\"m.room.member\",\"$ME\"],[\"m.room.canonical_alias\",\"\"],[\"m.room.name\",\"\"],[\"m.room.power_levels\",\"\"]],\"timeline_limit\":16}},\"extensions\":{\"account_data\":{\"enabled\":true}}}" path="/_matrix/client/unstable/org.matrix.msc3575/sync?timeout=3000" status=200 response_size="6.4 kiB"}: matrix_sdk::http_client: crates/matrix-sdk/src/http_client/mod.rs:204: Got response
2023-07-13T13:27:54.113694Z DEBUG root:sync_stream:sync_once: matrix_sdk::sliding_sync: crates/matrix-sdk/src/sliding_sync/mod.rs:542: Sliding Sync response received
2023-07-13T13:27:54.113790Z DEBUG root:sync_stream:sync_once: matrix_sdk::sliding_sync: crates/matrix-sdk/src/sliding_sync/mod.rs:558: Sliding Sync response handling starts
2023-07-13T13:27:54.113835Z  INFO root:sync_stream:sync_once:handle_response:process_sliding_sync:process_sliding_sync: matrix_sdk_base::sliding_sync: crates/matrix-sdk-base/src/sliding_sync.rs:65: rooms=0 lists=0 extensions=true
2023-07-13T13:27:54.113845Z  INFO root:sync_stream:sync_once:handle_response:process_sliding_sync:process_sliding_sync: matrix_sdk_base::sliding_sync: crates/matrix-sdk-base/src/sliding_sync.rs:77: to_device_events=0 device_one_time_keys_count=0
2023-07-13T13:27:54.116060Z DEBUG root:sync_stream:sync_once:handle_response:process_sliding_sync:process_sliding_sync: matrix_sdk_base::sliding_sync: crates/matrix-sdk-base/src/sliding_sync.rs:169: ready to submit changes to store
2023-07-13T13:27:54.116152Z DEBUG root:sync_stream:sync_once:handle_response:process_sliding_sync:process_sliding_sync: matrix_sdk_base::sliding_sync: crates/matrix-sdk-base/src/sliding_sync.rs:173: applied changes
2023-07-13T13:27:54.116204Z DEBUG root:sync_stream:sync_once:handle_response:process_sliding_sync: matrix_sdk::sliding_sync::client: crates/matrix-sdk/src/sliding_sync/client.rs:21: done processing on base_client
2023-07-13T13:27:54.116358Z DEBUG root:sync_stream:sync_once:handle_response: matrix_sdk::sliding_sync: crates/matrix-sdk/src/sliding_sync/mod.rs:279: Sliding Sync response has been handled by the client sync_response=SyncResponse { rooms: Rooms { leave: {}, join: {}, invite: {} }, account_data: [RawEvent { event_type: "io.element.recent_emoji", .. }, RawEvent { event_type: "org.matrix.msc3890.local_notification_settings.VSAYWBXPRS", .. }, RawEvent { event_type: "m.ignored_user_list", .. }, RawEvent { event_type: "m.cross_signing.master", .. }, RawEvent { event_type: "org.matrix.msc3890.local_notification_settings.NDARGFLGQK", .. }, RawEvent { event_type: "im.vector.setting.breadcrumbs", .. }, RawEvent { event_type: "m.push_rules", .. }, RawEvent { event_type: "m.secret_storage.key.35et9hHrx7M8lrtPXFSuaDuoqkzk9cY4", .. }, RawEvent { event_type: "m.secret_storage.default_key", .. }, RawEvent { event_type: "m.cross_signing.self_signing", .. }, RawEvent { event_type: "m.cross_signing.user_signing", .. }, RawEvent { event_type: "m.direct", .. }, RawEvent { event_type: "m.megolm_backup.v1", .. }, RawEvent { event_type: "im.vector.web.settings", .. }], to_device: [], device_lists: DeviceLists { changed: [], left: [] }, device_one_time_keys_count: {}, ambiguity_changes: AmbiguityChanges { changes: {} }, notifications: {}, .. }
2023-07-13T13:27:54.116431Z DEBUG root:sync_stream:sync_once:handle_response: matrix_sdk::sliding_sync: crates/matrix-sdk/src/sliding_sync/mod.rs:282: Update position markers` pos="1" delta_token=None
2023-07-13T13:27:54.116444Z DEBUG root:sync_stream:sync_once:handle_response: matrix_sdk::sliding_sync: crates/matrix-sdk/src/sliding_sync/mod.rs:347: Update lists lists={}
2023-07-13T13:27:54.116470Z TRACE root:sync_stream:sync_once: matrix_sdk::sliding_sync::cache: crates/matrix-sdk/src/sliding_sync/cache.rs:68: Saving a `SlidingSync` storage_key="sliding_sync_store::notifications::@mauro.devtest:matrix.org"
2023-07-13T13:27:54.116492Z DEBUG root:sync_stream:sync_once: matrix_sdk::sliding_sync: crates/matrix-sdk/src/sliding_sync/mod.rs:584: Sliding Sync response has been fully handled
2023-07-13T13:27:54.116545Z  WARN root: matrix_sdk_ui::notification_client: crates/matrix-sdk-ui/src/notification_client.rs:256: notification sync has run into an error: the sliding sync response doesn't include the target room
2023-07-13T13:27:54.117642Z  INFO root: nse: NotificationServiceExtension.swift:77: [NSE][0x0000000153d56210][0x0000000155a908d0] no notification for the event, discard
2023-07-13T13:27:54.118170Z  INFO root: nse: NotificationServiceExtension.swift:119: [NSE][0x0000000153d56210][0x0000000155a908d0] discard
2023-07-13T13:27:54.121410Z  INFO root: nse: NSELogger.swift:88: [NSE][0x0000000153d56210][0x0000000153e254b0] Memory: footprint: 16.45 MB - available: 7.55 MB
2023-07-13T13:27:54.122033Z  INFO root: nse: NotificationServiceExtension.swift:137: [NSE][0x0000000153d56210][0x0000000153e254b0] deinit

```

**Comment by bnjbvr** [comment]
Thanks @Velin92 for testing! Turns out this crash was unrelated, but we spotted some other issues and improvements that I'll add as todos in the initial comment as a personal reminder.

File: bindings/matrix-sdk-ffi/src/notification.rs [hunk 1]
```diff
@@ -3,11 +3,22 @@ use std::sync::Arc;
...
+    NotificationItem as MatrixNotificationItem,
 };
 use ruma::{EventId, RoomId};
 
 use crate::{error::ClientError, event::TimelineEvent, helpers::unwrap_or_clone_arc, RUNTIME};
 
+#[derive(uniffi::Enum)]
+pub enum NotificationEvent {
+    Timeline {
+        event: Arc<TimelineEvent>,
```
Comment (Velin92) [inline bindings/matrix-sdk-ffi/src/notification.rs:15]: I would go as far as saying that instead of returning `TimelineEvent` directly we could remove this type altogether, and just have the Timeline directly wrap the `TimelineEventType`

File: bindings/matrix-sdk-ffi/src/notification.rs [hunk 2]
```diff
@@ -3,11 +3,22 @@ use std::sync::Arc;
...
+    },
+    /// The notification is for an invitation; the interesting context bits live
+    /// in `NotificationItem`.
+    Invite,
+}
+
```
[the remark this answers is not in the document]
Comment (Velin92) [reply bindings/matrix-sdk-ffi/src/notification.rs:21]: Actually disregard what I said here, technically we already have the `event_id`.
The `sender_id` is in the `SenderInfo`
The timestamp could be useful but we are not using it as of right now.

File: crates/matrix-sdk-ui/src/notification_client.rs [hunk 3]
```diff
@@ -84,94 +94,317 @@ impl NotificationClient {
...
+        let maybe_event = notification.lock().unwrap().take();
+        Ok(maybe_event)
+    }
+
+    /// Get a full notification, given a room id and event id.
+    ///
+    /// This will run a small sliding sync to retrieve the content of the event,
+    /// along with extra data to form a rich notification context.
+    ///
+    /// For this to work, this sliding sync must not interfere with other
```
Comment (poljar) [inline crates/matrix-sdk-ui/src/notification_client.rs:278]: Doesn't the notification client need to use an in-memory store even if it's in the same process? Can't we enforce that the `NotificationClient` uses the memory store? I think we should be able to do so, no?

File: crates/matrix-sdk-ui/src/notification_client.rs [hunk 4]
```diff
@@ -84,94 +94,317 @@ impl NotificationClient {
...
+    /// same process, so that most of the contextual information for the
+    /// notification should already be there.
+    pub async fn get_notification_with_context(
+        &self,
+        room_id: &RoomId,
+        event_id: &EventId,
+    ) -> Result<Option<NotificationItem>, Error> {
+        tracing::info!("fetching notification event with a /context query");
+
+        // XXX Will this work for invites? A sliding sync may be running in the
```
Comment (poljar) [inline crates/matrix-sdk-ui/src/notification_client.rs:379]: Yeah, that's racy. Like discussed in the Matrix room, this should be called once a `SS` response has put the room into the memory store in the case the `SS` response didn't contain the event.

But let's do that in a separate PR and once the clients confirm that this is all fine and well.

**Review by poljar — APPROVED** [review]
I think this looks all fine, thanks for further clarifying what's going on in the Matrix room.

I think we're all on the same page where to go from here, left some minor comments but should be good to go.

File: bindings/matrix-sdk-ffi/src/notification.rs (hunk 1 above)
Comment (Velin92) [reply bindings/matrix-sdk-ffi/src/notification.rs]: oh no I don't mean changing the enum, I just meant that TimelineEvent is probably now not useful anymore so Timeline could be like this:
```
Timeline {
  event: Arc<TimelineEventType>
}
```

Since the only reason to have TimelineEvent was to wrap the TimelineEventType and expose in a generic way the `event_id `and the `sender_id` which now we have already either directly in app or in the `SenderInfo`

Comment (bnjbvr) [reply bindings/matrix-sdk-ffi/src/notification.rs]: Ah, that sounds about right; there's just the timestamp that might be useful, at least it's been used on Android as @jmartinesp told me earlier. It could be optional in one of the metadata structures, or filled somewhere else.

Comment (jmartinesp) [reply bindings/matrix-sdk-ffi/src/notification.rs]: The only values we're taking from `TimelineEvent` in Android are `senderId` and `eventId`, but we already have those  in the push notification so we can just pass them to our mapper class instead.

File: crates/matrix-sdk-ui/src/notification_client.rs (hunk 3 above)
Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/notification_client.rs:281]: That's now done, by inheriting the parent `Client`'s configuration into one that uses an in-memory state store in place of the one configured by the parent :+1:

File: bindings/matrix-sdk-ffi/src/notification.rs (hunk 1 above)
Comment (bnjbvr) [reply bindings/matrix-sdk-ffi/src/notification.rs]: Cool; I'm happy to change this in a followup PR, since this one is already bulky. Added that to #2330
